File: mobile_apps/ALADDIN_Android_New/app/src/main/java/family/aladdin/android/viewmodels/FamilyRegistrationViewModel.kt
package family.aladdin.android.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import family.aladdin.android.models.FamilyRole
import family.aladdin.android.models.AgeGroup
import family.aladdin.android.models.FamilyMember
import family.aladdin.android.models.CreateFamilyRequest
import family.aladdin.android.models.CreateFamilyResponse
import family.aladdin.android.models.JoinFamilyRequest
import family.aladdin.android.models.JoinFamilyResponse
import family.aladdin.android.models.RecoverFamilyResponse
import family.aladdin.android.models.FamilyMemberDetail
import family.aladdin.android.network.ApiService
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * 🏠 Family Registration ViewModel
 * Управление процессом прогрессивной регистрации
 * 
 * Отвечает за:
 * - Создание новой семьи
 * - Присоединение к существующей семье
 * - Восстановление доступа
 * - Интеграция с family_registration.py backend
 */

class FamilyRegistrationViewModel(
    private val apiService: ApiService,
    private val context: android.content.Context
) : ViewModel() {
    
    // MARK: - State
    
    private val _currentStep = MutableStateFlow(RegistrationStep.IDLE)
    val currentStep: StateFlow<RegistrationStep> = _currentStep.asStateFlow()
    
    private val _showConsentModal = MutableStateFlow(false)
    val showConsentModal: StateFlow<Boolean> = _showConsentModal.asStateFlow()
    
    private val _showRoleModal = MutableStateFlow(false)
    val showRoleModal: StateFlow<Boolean> = _showRoleModal.asStateFlow()
    
    private val _showAgeGroupModal = MutableStateFlow(false)
    val showAgeGroupModal: StateFlow<Boolean> = _showAgeGroupModal.asStateFlow()
    
    private val _showLetterModal = MutableStateFlow(false)
    val showLetterModal: StateFlow<Boolean> = _showLetterModal.asStateFlow()
    
    private val _showFamilyCreatedModal = MutableStateFlow(false)
    val showFamilyCreatedModal: StateFlow<Boolean> = _showFamilyCreatedModal.asStateFlow()
    
    private val _showSuccessModal = MutableStateFlow(false)
    val showSuccessModal: StateFlow<Boolean> = _showSuccessModal.asStateFlow()
    
    private val _consentAccepted = MutableStateFlow(false)
    val consentAccepted: StateFlow<Boolean> = _consentAccepted.asStateFlow()
    
    private val _selectedRole = MutableStateFlow<FamilyRole?>(null)
    val selectedRole: StateFlow<FamilyRole?> = _selectedRole.asStateFlow()
    
    private val _selectedAgeGroup = MutableStateFlow<AgeGroup?>(null)
    val selectedAgeGroup: StateFlow<AgeGroup?> = _selectedAgeGroup.asStateFlow()
    
    private val _selectedLetter = MutableStateFlow<String?>(null)
    val selectedLetter: StateFlow<String?> = _selectedLetter.asStateFlow()
    
    private val _familyID = MutableStateFlow<String?>(null)
    val familyID: StateFlow<String?> = _familyID.asStateFlow()
    
    private val _recoveryCode = MutableStateFlow<String?>(null)
    val recoveryCode: StateFlow<String?> = _recoveryCode.asStateFlow()
    
    private val _familyMembers = MutableStateFlow<List<FamilyMember>>(emptyList())
    val familyMembers: StateFlow<List<FamilyMember>> = _familyMembers.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    // MARK: - Registration Steps
    
    enum class RegistrationStep {
        IDLE,
        SHOWING_CONSENT,
        SELECTING_ROLE,
        SELECTING_AGE_GROUP,
        SELECTING_LETTER,
        CREATING_FAMILY,
        SHOWING_RECOVERY_CODE,
        COMPLETED
    }
    
    // MARK: - Public Methods
    
    /**
     * Начать регистрацию (показать Consent)
     */
    fun startRegistration() {
        viewModelScope.launch {
            // Проверяем, было ли согласие дано ранее
            val sharedPrefs = context.getSharedPreferences("aladdin_prefs", android.content.Context.MODE_PRIVATE)
            
            if (sharedPrefs.getBoolean("consent_accepted", false)) {
                // Согласие уже дано - пропускаем
                _consentAccepted.value = true
                showRoleSelection()
            } else {
                // Показываем Consent Modal
                _currentStep.value = RegistrationStep.SHOWING_CONSENT
                _showConsentModal.value = true
            }
        }
    }
    
    /**
     * Согласие принято
     */
    fun acceptConsent() {
        viewModelScope.launch {
            _consentAccepted.value = true
            
            // Сохраняем в SharedPreferences
            val sharedPrefs = context.getSharedPreferences("aladdin_prefs", android.content.Context.MODE_PRIVATE)
            sharedPrefs.edit().apply {
                putBoolean("consent_accepted", true)
                putLong("consent_date", System.currentTimeMillis())
                putString("consent_version", "2.0")
                apply()
            }
            
            _showConsentModal.value = false
            
            // Переходим к выбору роли
            showRoleSelection()
        }
    }
    
    /**
     * Показать выбор роли
     */
    fun showRoleSelection() {
        _currentStep.value = RegistrationStep.SELECTING_ROLE
        _showRoleModal.value = true
    }
    
    /**
     * Выбрана роль
     */
    fun onRoleSelected(role: FamilyRole) {
        viewModelScope.launch {
            _selectedRole.value = role
            _showRoleModal.value = false
            
            delay(500)
            
            _currentStep.value = RegistrationStep.SELECTING_AGE_GROUP
            _showAgeGroupModal.value = true
        }
    }
    
    fun onAgeGroupSelected(ageGroup: AgeGroup) {
        viewModelScope.launch {
            _selectedAgeGroup.value = ageGroup
            _showAgeGroupModal.value = false
            
            delay(500)
            
            _currentStep.value = RegistrationStep.SELECTING_LETTER
            _showLetterModal.value = true
        }
    }
    
    fun onLetterSelected(letter: String) {
        viewModelScope.launch {
            _selectedLetter.value = letter
            _showLetterModal.value = false
            
            delay(500)
            
            createFamily()
        }
    }
    
    // MARK: - Create Family
    
    private fun createFamily() {
        val role = _selectedRole.value ?: return
        val ageGroup = _selectedAgeGroup.value ?: return
        val letter = _selectedLetter.value ?: return
        
        viewModelScope.launch {
            _currentStep.value = RegistrationStep.CREATING_FAMILY
            _isLoading.value = true
            
            try {
                val request = CreateFamilyRequest(
                    role = role.value,
                    age_group = ageGroup.value,
                    personal_letter = letter,
                    device_type = getDeviceType()
                )
                
                val response = apiService.createFamily(request)
                
                _familyID.value = response.family_id
                _recoveryCode.value = response.recovery_code
                
                _currentStep.value = RegistrationStep.SHOWING_RECOVERY_CODE
                
                delay(500)
                
                _showFamilyCreatedModal.value = true
                
            } catch (e: Exception) {
                _errorMessage.value = e.message
                android.util.Log.e("FamilyRegistration", "Error creating family", e)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    // MARK: - Join Family
    
    fun joinFamily(familyCode: String) {
        val role = _selectedRole.value ?: return
        val ageGroup = _selectedAgeGroup.value ?: return
        val letter = _selectedLetter.value ?: return
        
        viewModelScope.launch {
            _isLoading.value = true
            
            try {
                val request = JoinFamilyRequest(
                    family_id = extractFamilyID(familyCode),
                    role = role.value,
                    age_group = ageGroup.value,
                    personal_letter = letter,
                    device_type = getDeviceType()
                )
                
                val response = apiService.joinFamily(request)
                
                _familyID.value = response.family_id
                _familyMembers.value = response.members.map { member ->
                    FamilyMember(
                        id = member.member_id,
                        letter = member.personal_letter,
                        role = FamilyRole.values().find { it.value == member.role } ?: FamilyRole.OTHER,
                        ageGroup = AgeGroup.values().find { it.value == member.age_group } ?: AgeGroup.ADULT_24_55,
                        isYou = member.member_id == response.your_member_id
                    )
                }
                
                _currentStep.value = RegistrationStep.COMPLETED
                _showSuccessModal.value = true
                
            } catch (e: Exception) {
                _errorMessage.value = e.message
                android.util.Log.e("FamilyRegistration", "Error joining family", e)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    // MARK: - Recover Access
    
    fun recoverAccess(recoveryCode: String) {
        viewModelScope.launch {
            _isLoading.value = true
            
            try {
                val familyId = extractFamilyID(recoveryCode)
                val response = apiService.recoverFamily(familyId)
                
                _familyID.value = response.family_id
                _familyMembers.value = response.members.map { member ->
                    FamilyMember(
                        id = member.member_id,
                        letter = member.personal_letter,
                        role = FamilyRole.values().find { it.value == member.role } ?: FamilyRole.OTHER,
                        ageGroup = AgeGroup.values().find { it.value == member.age_group } ?: AgeGroup.ADULT_24_55,
                        isYou = false
                    )
                }
                
                _currentStep.value = RegistrationStep.COMPLETED
                _showSuccessModal.value = true
                
            } catch (e: Exception) {
                _errorMessage.value = e.message
                android.util.Log.e("FamilyRegistration", "Error recovering family", e)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    // MARK: - Helper Methods
    
    private fun extractFamilyID(code: String): String {
        // Convert FAM-A1B2-C3D4-E5F6 → FAM_A1B2C3D4E5F6
        return code.replace("-", "").replace("FAM", "FAM_")
    }
    
    private fun getDeviceType(): String {
        return "smartphone"  // TODO: Detect tablet
    }
    
    /**
     * Закрыть модалку "Семья создана"
     */
    fun dismissFamilyCreatedModal() {
        _showFamilyCreatedModal.value = false
    }
    
    /**
     * Закрыть модалку "Успешная регистрация"
     */
    fun dismissSuccessModal() {
        _showSuccessModal.value = false
    }
}

// MARK: - API Service Extension

suspend fun ApiService.createFamily(request: CreateFamilyRequest): CreateFamilyResponse {
    // TODO: Implement actual API call using request
    return CreateFamilyResponse(
        success = true,
        family_id = "FAM_A1B2C3D4E5F6",
        recovery_code = "FAM-A1B2-C3D4-E5F6",
        qr_code_data = "{}",
        short_code = "AB12",
        member_id = "MEM_X1Y2Z3W4"
    )
}

suspend fun ApiService.joinFamily(request: JoinFamilyRequest): JoinFamilyResponse {
    // TODO: Implement actual API call
    return JoinFamilyResponse(
        success = true,
        family_id = request.family_id,
        your_member_id = "MEM_NEW123",
        role = request.role,
        personal_letter = request.personal_letter,
        members = emptyList()
    )
}

suspend fun ApiService.recoverFamily(familyId: String): RecoverFamilyResponse {
    // TODO: Implement actual API call
    return RecoverFamilyResponse(
        success = true,
        family_id = familyId,
        members = emptyList(),
        family_status = emptyMap()
    )
}

File: ALADDIN_NEW/mobile_apps/ALADDIN_iOS/ViewModels/FamilyRegistrationViewModel.swift
import SwiftUI
import Combine

/**
 * 🏠 Family Registration ViewModel
 * Управление процессом прогрессивной регистрации
 * 
 * Отвечает за:
 * - Создание новой семьи
 * - Присоединение к существующей семье
 * - Восстановление доступа
 * - Интеграция с family_registration.py backend
 */

class FamilyRegistrationViewModel: ObservableObject {
    
    // MARK: - Published Properties
    
    @Published var currentStep: RegistrationStep = .idle
    @Published var showConsentModal: Bool = false
    @Published var showRoleModal: Bool = false
    @Published var showAgeGroupModal: Bool = false
    @Published var showLetterModal: Bool = false
    @Published var showFamilyCreatedModal: Bool = false
    @Published var showSuccessModal: Bool = false
    @Published var consentAccepted: Bool = false
    
    @Published var selectedRole: FamilyRole?
    @Published var selectedAgeGroup: AgeGroup?
    @Published var selectedLetter: String?
    
    @Published var familyID: String?
    @Published var recoveryCode: String?
    @Published var familyMembers: [FamilyMember] = []
    
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    // MARK: - Registration Steps
    
    enum RegistrationStep {
        case idle
        case showingConsent
        case selectingRole
        case selectingAgeGroup
        case selectingLetter
        case creatingFamily
        case showingRecoveryCode
        case completed
    }
    
    // MARK: - Public Methods
    
    /**
     * Начать регистрацию (показать Consent)
     */
    func startRegistration() {
        // Проверяем, было ли согласие дано ранее
        if UserDefaults.standard.bool(forKey: "consent_accepted") {
            // Согласие уже дано - пропускаем
            consentAccepted = true
            showRoleSelection()
        } else {
            // Показываем Consent Modal
            currentStep = .showingConsent
            showConsentModal = true
        }
    }
    
    /**
     * Согласие принято
     */
    func acceptConsent() {
        consentAccepted = true
        UserDefaults.standard.set(true, forKey: "consent_accepted")
        UserDefaults.standard.set(Date(), forKey: "consent_date")
        UserDefaults.standard.set("2.0", forKey: "consent_version")
        showConsentModal = false
        
        // Переходим к выбору роли
        showRoleSelection()
    }
    
    /**
     * Показать выбор роли
     */
    func showRoleSelection() {
        currentStep = .selectingRole
        showRoleModal = true
    }
    
    func onRoleSelected(_ role: FamilyRole) {
        selectedRole = role
        showRoleModal = false
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.currentStep = .selectingAgeGroup
            self.showAgeGroupModal = true
        }
    }
    
    func onAgeGroupSelected(_ ageGroup: AgeGroup) {
        selectedAgeGroup = ageGroup
        showAgeGroupModal = false
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.currentStep = .selectingLetter
            self.showLetterModal = true
        }
    }
    
    func onLetterSelected(_ letter: String) {
        selectedLetter = letter
        showLetterModal = false
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.createFamily()
        }
    }
    
    // MARK: - Create Family
    
    func createFamily() {
        guard let role = selectedRole,
              let ageGroup = selectedAgeGroup,
              let letter = selectedLetter else {
            return
        }
        
        currentStep = .creatingFamily
        isLoading = true
        
        // API request
        let request = CreateFamilyRequest(
            role: role.rawValue,
            age_group: ageGroup.rawValue,
            personal_letter: letter,
            device_type: getDeviceType()
        )
        
        NetworkManager.shared.createFamily(request: request) { [weak self] result in
            DispatchQueue.main.async {
                self?.isLoading = false
                
                switch result {
                case .success(let response):
                    self?.familyID = response.family_id
                    self?.recoveryCode = response.recovery_code
                    
                    // Format recovery code for display
                    self?.recoveryCode = self?.formatRecoveryCode(response.family_id) ?? response.recovery_code
                    
                    self?.currentStep = .showingRecoveryCode
                    
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                        self?.showFamilyCreatedModal = true
                    }
                    
                case .failure(let error):
                    self?.errorMessage = error.localizedDescription
                    print("❌ Error creating family: \(error)")
                }
            }
        }
    }
    
    // MARK: - Join Family
    
    func joinFamily(withCode code: String) {
        guard let role = selectedRole,
              let ageGroup = selectedAgeGroup,
              let letter = selectedLetter else {
            return
        }
        
        isLoading = true
        
        let request = JoinFamilyRequest(
            family_id: extractFamilyID(from: code),
            role: role.rawValue,
            age_group: ageGroup.rawValue,
            personal_letter: letter,
            device_type: getDeviceType()
        )
        
        NetworkManager.shared.joinFamily(request: request) { [weak self] result in
            DispatchQueue.main.async {
                self?.isLoading = false
                
                switch result {
                case .success(let response):
                    self?.familyID = response.family_id
                    self?.familyMembers = response.members.map { member in
                        FamilyMember(
                            id: member.member_id,
                            letter: member.personal_letter,
                            role: FamilyRole(rawValue: member.role) ?? .other,
                            ageGroup: AgeGroup(rawValue: member.age_group) ?? .adult_24_55,
                            isYou: member.member_id == response.your_member_id
                        )
                    }
                    
                    self?.currentStep = .completed
                    self?.showSuccessModal = true
                    
                case .failure(let error):
                    self?.errorMessage = error.localizedDescription
                    print("❌ Error joining family: \(error)")
                }
            }
        }
    }
    
    // MARK: - Recover Access
    
    func recoverAccess(withCode code: String) {
        isLoading = true
        
        NetworkManager.shared.recoverFamily(familyID: extractFamilyID(from: code)) { [weak self] result in
            DispatchQueue.main.async {
                self?.isLoading = false
                
                switch result {
                case .success(let response):
                    self?.familyID = response.family_id
                    self?.familyMembers = response.members.map { member in
                        FamilyMember(
                            id: member.member_id,
                            letter: member.personal_letter,
                            role: FamilyRole(rawValue: member.role) ?? .other,
                            ageGroup: AgeGroup(rawValue: member.age_group) ?? .adult_24_55,
                            isYou: false
                        )
                    }
                    
                    self?.currentStep = .completed
                    self?.showSuccessModal = true
                    
                case .failure(let error):
                    self?.errorMessage = error.localizedDescription
                    print("❌ Error recovering family: \(error)")
                }
            }
        }
    }
    
    // MARK: - Helper Methods
    
    private func formatRecoveryCode(_ familyID: String) -> String {
        // Convert FAM_A1B2C3D4E5F6 → FAM-A1B2-C3D4-E5F6
        let cleaned = familyID.replacingOccurrences(of: "FAM_", with: "")
        let parts = cleaned.enumerated().reduce(into: [String]()) { result, element in
            let index = element.offset
            let char = element.element
            
            if index % 4 == 0 {
                result.append(String(char))
            } else {
                result[result.count - 1].append(char)
            }
        }
        
        return "FAM-" + parts.joined(separator: "-")
    }
    
    private func extractFamilyID(from code: String) -> String {
        // Convert FAM-A1B2-C3D4-E5F6 → FAM_A1B2C3D4E5F6
        let cleaned = code.replacingOccurrences(of: "-", with: "")
        return cleaned.replacingOccurrences(of: "FAM", with: "FAM_")
    }
    
    private func getDeviceType() -> String {
        #if os(iOS)
        if UIDevice.current.userInterfaceIdiom == .pad {
            return "tablet"
        } else {
            return "smartphone"
        }
        #else
        return "smartphone"
        #endif
    }
}

// MARK: - API Models

struct CreateFamilyRequest: Codable {
    let role: String
    let age_group: String
    let personal_letter: String
    let device_type: String
}

struct CreateFamilyResponse: Codable {
    let family_id: String
    let recovery_code: String
    let qr_code_data: String
    let short_code: String
}

struct JoinFamilyRequest: Codable {
    let family_id: String
    let role: String
    let age_group: String
    let personal_letter: String
    let device_type: String
}

struct JoinFamilyResponse: Codable {
    let family_id: String
    let your_member_id: String
    let members: [FamilyMemberResponse]
}

struct FamilyMemberResponse: Codable {
    let member_id: String
    let role: String
    let age_group: String
    let personal_letter: String
}

struct RecoverFamilyResponse: Codable {
    let family_id: String
    let members: [FamilyMemberResponse]
}

// MARK: - Network Manager Extension

extension NetworkManager {
    
    func createFamily(request: CreateFamilyRequest, completion: @escaping (Result<CreateFamilyResponse, Error>) -> Void) {
        guard let url = URL(string: "\(AppConfig.baseURL)/family/create") else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1)))
            return
        }
        
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try? JSONEncoder().encode(request)
        
        URLSession.shared.dataTask(with: urlRequest) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data", code: -1)))
                return
            }
            
            do {
                let decoded = try JSONDecoder().decode(CreateFamilyResponse.self, from: data)
                completion(.success(decoded))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    func joinFamily(request: JoinFamilyRequest, completion: @escaping (Result<JoinFamilyResponse, Error>) -> Void) {
        guard let url = URL(string: "\(AppConfig.baseURL)/family/join") else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1)))
            return
        }
        
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try? JSONEncoder().encode(request)
        
        URLSession.shared.dataTask(with: urlRequest) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data", code: -1)))
                return
            }
            
            do {
                let decoded = try JSONDecoder().decode(JoinFamilyResponse.self, from: data)
                completion(.success(decoded))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    func recoverFamily(familyID: String, completion: @escaping (Result<RecoverFamilyResponse, Error>) -> Void) {
        guard let url = URL(string: "\(AppConfig.baseURL)/family/recover/\(familyID)") else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1)))
            return
        }
        
        URLSession.shared.dataTask(with: url) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data", code: -1)))
                return
            }
            
            do {
                let decoded = try JSONDecoder().decode(RecoverFamilyResponse.self, from: data)
                completion(.success(decoded))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
}

